server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/UserHeightCalibration.kt

```diff
@@ -0,0 +1,317 @@
+	val dotProduct = worldHmdUp.dot(Vector3.POS_Y)
+	return dotProduct >= threshold
+}
+
+fun isControllerPointingDown(controller: Tracker, threshold: Double): Boolean {
+	val q = controller.getRotation()
+	val controllerForwardWorld = q.sandwich(Vector3.POS_Z)
+	val worldDown = Vector3.NEG_Y
+	val dotProduct = controllerForwardWorld.dot(worldDown)
+
+	return dotProduct >= threshold
+}
+
+interface UserHeightCalibrationListener {
+	fun onStatusChange(status: UserHeightRecordingStatusResponseT)
+}
+
+class UserHeightCalibration(val server: VRServer, val humanPoseManager: HumanPoseManager) {
+	var status = UserHeightCalibrationStatus.NONE
+
+	var currentHeight = 0f
+	var currentFloorLevel = 0f
+
+	var startTime = 0f
+
+	private val hmdPositionSamples: CircularFifoQueue<Vector3> = CircularFifoQueue(MAX_SAMPLES)
+	private var heightStableStartTime: Float? = null
+
+	private val floorPositionSamples: CircularFifoQueue<Vector3> = CircularFifoQueue(MAX_SAMPLES)
+	private var floorStableStartTime: Float? = null
+
+	private val listeners: MutableList<UserHeightCalibrationListener> = CopyOnWriteArrayList()
+	private var hmd: Tracker? = null
+
+	private val handTrackers: MutableList<Tracker> = mutableListOf()
+
+	fun start() {
+		clear()
+		checkTrackers()
+
+		if (!server.serverGuards.canDoUserHeightCalibration) {
+			return
+		}
+
+		startTime = System.nanoTime().toFloat()
+
+		status = UserHeightCalibrationStatus.RECORDING_FLOOR
+		currentFloorLevel = Float.MAX_VALUE
+
+		sendStatusUpdate()
+	}
+
+	fun clear() {
+		status = UserHeightCalibrationStatus.NONE
+		hmdPositionSamples.clear()
+		floorPositionSamples.clear()
+		heightStableStartTime = null
+		floorStableStartTime = null
+		currentHeight = 0f
+		startTime = 0f
+	}
+
+	init {
+		clear()
+		checkTrackers()
+	}
+
+	fun checkTrackers() {
+		handTrackers.clear()
+		handTrackers.addAll(
+			server.allTrackers.filter {
+				(
+					it.trackerPosition == TrackerPosition.LEFT_HAND ||
+						it.trackerPosition == TrackerPosition.RIGHT_HAND
+					) &&
+					!it.isInternal &&
+					it.hasPosition &&
+					it.status == TrackerStatus.OK
+			},
+		)
+
+		hmd = server.allTrackers.find {
+			it.trackerPosition == TrackerPosition.HEAD &&
+				it.hasPosition &&
+				!it.isInternal &&
+				it.status == TrackerStatus.OK
+		}
+		server.serverGuards.canDoUserHeightCalibration = hmd != null && handTrackers.isNotEmpty()
+
+		currentHeight = 0f
+		currentFloorLevel = 0f
+	}
+
+	fun applyCalibration() {
+		server.configManager.vrConfig.skeleton.hmdHeight = currentHeight
+		server.configManager.vrConfig.skeleton.floorHeight = 0f
+
+		server.humanPoseManager.resetOffsets()
+		server.humanPoseManager.saveConfig()
+		server.configManager.saveConfig()
+
+		server.trackingChecklistManager.resetMountingCompleted = false
+		server.trackingChecklistManager.feetResetMountingCompleted = false
+	}
+
+	fun tick() {
+		if (startTime == 0f) return
+
+		val currentTime = System.nanoTime().toFloat()
```

Comment:
Normally we want to use our own time tracking within tick code (where precision doesn't matter) so we can unit test more easily (ex `tickInterval` variable or smth that is accumulated over time), but it doesn't really matter right now. I do gotta wonder why on earth this is a float though? We're losing our long precision for no reason.